# gatk-snp-xpclr
author: caominghao    date: 2024.6.26
****
## 1.准备数据
```shell
ln -s /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr*.vcf.gz ./
```
## 2.基于各染色体的vcf文件生成tped文件
```shell
# 生成list
less early.list | awk '{print $1 "\t" $1}' > early.keep.list
less late.list | awk '{print $1 "\t" $1}' > late.keep.list

# 生成tped
## early
for i in {1..18};
do
plink --vcf /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr${i}.vcf.gz --keep /home/caominghao/02.HJB/06.pop_select/early.keep.list --out /home/caominghao/02.HJB/06.pop_select/03.XPCLR/tped/early.Chr${i} --recode 01 transpose -output-missing-genotype 9 --allow-extra-chr --keep-allele-order --double-id & 
done
## late
for i in {1..18};
do
plink --vcf /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr${i}.vcf.gz --keep /home/caominghao/02.HJB/06.pop_select/late.keep.list --out /home/caominghao/02.HJB/06.pop_select/03.XPCLR/tped/late.Chr${i} --recode 01 transpose -output-missing-genotype 9 --allow-extra-chr --keep-allele-order --double-id & 
done
```
## 3.基于tped文件生成geno文件(比较快)
```shell
# early
for i in {1..18};
do
cut -d " " -f 5- /home/caominghao/02.HJB/06.pop_select/03.XPCLR/tped/early.Chr${i}.tped | awk '{print $0" "}' > /home/caominghao/02.HJB/06.pop_select/03.XPCLR/geno/early.Chr${i}.geno & 
done
# late
for i in {1..18};
do
cut -d " " -f 5- /home/caominghao/02.HJB/06.pop_select/03.XPCLR/tped/late.Chr${i}.tped | awk '{print $0" "}' > /home/caominghao/02.HJB/06.pop_select/03.XPCLR/geno/late.Chr${i}.geno & 
done
```
## 4.基于vcf文件生成snp文件
```shell

## 输出文件路径
/home/caominghao/02.HJB/06.pop_select/


zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr1.vcf.gz |awk '$1=="1" {print " "$1":"$2 "\tChr1\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr1.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr2.vcf.gz |awk '$1=="2" {print " "$1":"$2 "\tChr2\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr2.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr3.vcf.gz |awk '$1=="3" {print " "$1":"$2 "\tChr3\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr3.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr4.vcf.gz |awk '$1=="4" {print " "$1":"$2 "\tChr4\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr4.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr5.vcf.gz |awk '$1=="5" {print " "$1":"$2 "\tChr5\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr5.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr6.vcf.gz |awk '$1=="6" {print " "$1":"$2 "\tChr6\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr6.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr7.vcf.gz |awk '$1=="7" {print " "$1":"$2 "\tChr7\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr7.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr8.vcf.gz |awk '$1=="8" {print " "$1":"$2 "\tChr8\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr8.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr9.vcf.gz |awk '$1=="9" {print " "$1":"$2 "\tChr9\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr9.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr10.vcf.gz |awk '$1=="10" {print " "$1":"$2 "\tChr10\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr10.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr11.vcf.gz |awk '$1=="11" {print " "$1":"$2 "\tChr11\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr11.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr12.vcf.gz |awk '$1=="12" {print " "$1":"$2 "\tChr12\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr12.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr13.vcf.gz |awk '$1=="13" {print " "$1":"$2 "\tChr13\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr13.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr14.vcf.gz |awk '$1=="14" {print " "$1":"$2 "\tChr14\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr14.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr15.vcf.gz |awk '$1=="15" {print " "$1":"$2 "\tChr15\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr15.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr16.vcf.gz |awk '$1=="16" {print " "$1":"$2 "\tChr16\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr16.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr17.vcf.gz |awk '$1=="17" {print " "$1":"$2 "\tChr17\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr17.snp
zcat /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr18.vcf.gz |awk '$1=="18" {print " "$1":"$2 "\tChr18\t" $2/100000000 "\t" $2 "\t"  $4 "\t" $5 }' >  /home/caominghao/02.HJB/06.pop_select/snp/493.snp.filter.pass.gt.noN.Chr18.snp
```
## 5.计算XP-CLR
```shell
# 18条染色体一起跑
for i in {1..18};do time xpclr --format txt --popA /home/caominghao/02.HJB/06.pop_select/03.XPCLR/geno/early.Chr${i}.geno --popB /home/caominghao/02.HJB/06.pop_select/03.XPCLR/geno/late.Chr${i}.geno --map /home/caominghao/02.HJB/06.pop_select/03.XPCLR/snp/493.snp.filter.pass.gt.noN.Chr${i}.snp --out /home/caominghao/02.HJB/06.pop_select/03.XPCLR/early-late.Chr${i}.out --chr ${i} --maxsnps 600 --size 50000 --step 25000 --phased & done
# 单独跑1号染色体
time xpclr --format txt --popA early.Chr10.geno --popB late.Chr10.geno --map ../snp/HJB.Chr10.snp --out ./early-late.Chr10.out --chr 10 --maxsnps 600 --size 50000 --step 25000 --phased
```
## 6.加表头
```shell
## 去表头
for i in {1..18};do sed -i '1d' early-late.Chr${i}.out;done
## 合并
cat early-late.Chr1.out early-late.Chr2.out early-late.Chr3.out early-late.Chr4.out early-late.Chr5.out early-late.Chr6.out early-late.Chr7.out early-late.Chr8.out early-late.Chr9.out early-late.Chr10.out early-late.Chr11.out early-late.Chr12.out early-late.Chr13.out early-late.Chr14.out early-late.Chr15.out early-late.Chr16.out early-late.Chr17.out early-late.Chr18.out > early-late.ChrAll.out
## 加表头
id      chrom   start   stop    pos_start       pos_stop        modelL  nullL   sel_coef        nSNPs   nSNPs_avail     xpclr   xpclr_norm
```
## 7.画图
```shell
awk '$12>0 {print $0}' early-late.ChrAll.out > early-late.ChrAll.qc.out
awk '$11>0 {print $0}' early-late.ChrAll.qc.out > early-late.ChrAll.qc2.out
python3 /home/caominghao/software/script/plot_Manhattan.py --infile early-late.ChrAll.out --chr-col chrom --loc-col start --val-col xpclr --outfile gatk-snp-xpclr.Manhattan.pdf --xlabel Chromosome --ylabel XP-CLR --figsize 30 9 --axlabelsize 24 --markersize 12 --ticklabelsize 24
```
## 8.提取5%的点并使用annovar注释
```shell

## 在/home/caominghao/02.HJB/06.pop_select/05.candid_gene/01.select_gene/03.XPCLR/处理XPCLR的结果:TOP5%的位点
#  转bed
sort -k 12 -g XPCLR.txt | tail -n 4125 | awk 'BEGIN {OFS="\t"} {print $2,$3,$4,$12}' > XPCLR.bed
# 注释
sed -i 's/^10/chr10/g' *bed
sed -i 's/^11/chr11/g' *bed
sed -i 's/^12/chr12/g' *bed
sed -i 's/^13/chr13/g' *bed
sed -i 's/^14/chr14/g' *bed
sed -i 's/^15/chr15/g' *bed
sed -i 's/^16/chr16/g' *bed
sed -i 's/^17/chr17/g' *bed
sed -i 's/^18/chr18/g' *bed
sed -i 's/^1/chr1/g' *bed
sed -i 's/^2/chr2/g' *bed
sed -i 's/^3/chr3/g' *bed
sed -i 's/^4/chr4/g' *bed
sed -i 's/^5/chr5/g' *bed
sed -i 's/^6/chr6/g' *bed
sed -i 's/^7/chr7/g' *bed
sed -i 's/^8/chr8/g' *bed
sed -i 's/^9/chr9/g' *bed
# 第三步（给bed文件添加表头）
#### 表头内容如下
track name="my_bed_file" description="This is a BED file"
# 第四步（将bed文件转为annovar输入文件）
perl /home/caominghao/software/21.ANNOVAR/annovar/convert2annovar.pl -format bed XPCLR.bed > XPCLR.avinput
# 注释
perl /home/caominghao/software/21.ANNOVAR/annovar/annotate_variation.pl -out XPCLR -build Duroc XPCLR.avinput /home/caominghao/software/21.ANNOVAR/annovar/Duroc/
```
## 5.提取基因
```shell
cat XPCLR.variant_function | awk -F "\t" '{print $2}' | grep -v ^NONE | awk -F ',' 'BEGIN {OFS="\n"} {print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25,$26,$27,$28,$29,$30}' | sort | uniq | grep -v ^rna- | grep -v ^NONE | grep -v TRNA | awk -F "(" '{print $1}' | awk -F "-" '{print $2}' | sort | uniq | less -SN
```

