# gatk-snp-pi
author: caominghao    date: 2024.6.26
****
## 1.准备文件
```shell
ln -s /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr*.vcf.gz ./
```
## 2.计算PI值
```shell
# 计算HJB群体18条染色体的Pi
for i in {1..18};
do
vcftools --gzvcf /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr${i}.vcf.gz \
--keep /home/caominghao/02.HJB/06.pop_select/early.list --window-pi 50000 --window-pi-step 25000 --out /home/caominghao/02.HJB/06.pop_select/02.PI/early.Chr${i} --max-missing 0.1 --maf 0.01 &
done



for i in {1..18};
do
vcftools --gzvcf /home/caominghao/02.HJB/01.vcf/snp/493.snp.filter.pass.gt.noN.Chr${i}.vcf.gz \
--keep /home/caominghao/02.HJB/06.pop_select/late.list --window-pi 50000 --window-pi-step 25000 --out /home/caominghao/02.HJB/06.pop_select/02.PI/late.Chr${i} --max-missing 0.1 --maf 0.01 &
done




# 整合两个群体的Pi文件
## 修改格式
for i in {1..18};do less early.Chr${i}.windowed.pi |awk '{print $1 "-" $2 "-" $3 "\t" $5}' > early.Chr${i}.changeFormat.windowed.pi;done
for i in {1..18};do less late.Chr${i}.windowed.pi |awk '{print $1 "-" $2 "-" $3 "\t" $5}' > late.Chr${i}.changeFormat.windowed.pi;done
## 去表头
for i in {1..18};do sed -i '1d' early.Chr${i}.changeFormat.windowed.pi;done
for i in {1..18};do sed -i '1d' late.Chr${i}.changeFormat.windowed.pi;done
for i in {1..18};do mv early.Chr${i}.changeFormat.windowed.pi early.Chr${i}.windowed.pi;done
for i in {1..18};do mv late.Chr${i}.changeFormat.windowed.pi late.Chr${i}.windowed.pi;done
## 合并
cat late.Chr1.windowed.pi late.Chr2.windowed.pi late.Chr3.windowed.pi late.Chr4.windowed.pi late.Chr5.windowed.pi late.Chr6.windowed.pi late.Chr7.windowed.pi late.Chr8.windowed.pi late.Chr9.windowed.pi late.Chr10.windowed.pi late.Chr11.windowed.pi late.Chr12.windowed.pi late.Chr13.windowed.pi late.Chr14.windowed.pi late.Chr15.windowed.pi late.Chr16.windowed.pi late.Chr17.windowed.pi late.Chr18.windowed.pi > late.pi
cat early.Chr1.windowed.pi early.Chr2.windowed.pi early.Chr3.windowed.pi early.Chr4.windowed.pi early.Chr5.windowed.pi early.Chr6.windowed.pi early.Chr7.windowed.pi early.Chr8.windowed.pi early.Chr9.windowed.pi early.Chr10.windowed.pi early.Chr11.windowed.pi early.Chr12.windowed.pi early.Chr13.windowed.pi early.Chr14.windowed.pi early.Chr15.windowed.pi early.Chr16.windowed.pi early.Chr17.windowed.pi early.Chr18.windowed.pi > early.pi
# 合并
csvtk join -t -H -f 1 early.pi late.pi > window.pi
# 除
less window.pi  | awk '{print $1 "\t" $2/$3}' > window.piratio
# # 将-变为Tab键
less window.piratio | awk -F- '{print $1 "\t" $2 "\t" $3}' > chr.piratio
```
## 3.可视化
```shell
# 
awk -F "\t" '$4>0 {print $0}' chr.piratio > chr.qc.piratio
awk '$4<100 {print $0}' chr.qc.piratio > chr.qc2.piratio

# 加表头
CHROM   BIN_START       BIN_END      PI


python3 /home/caominghao/software/script/plot_Manhattan.py \
--infile chr.qc2.piratio --chr-col CHROM --loc-col BIN_START --val-col PI \
--outfile gatk-snp-pi.Manhattan.pdf --xlabel Chromosome --ylabel πearly/πlate \
--figsize 30 9 --axlabelsize 24 --markersize 12 --ticklabelsize 24
```

## 4.提取5%的点并使用annovar注释

```shell

## 在/home/caominghao/02.HJB/06.pop_select/05.candid_gene/01.select_gene/02.PI/处理PI的结果:TOP5%的位点
#  转bed
sort -k 4 -g PI.txt | tail -n 3910 | awk 'BEGIN {OFS="\t"} {print $1,$2,$3,$4}' > PI.bed
# 注释
sed -i 's/^10/chr10/g' *bed
sed -i 's/^11/chr11/g' *bed
sed -i 's/^12/chr12/g' *bed
sed -i 's/^13/chr13/g' *bed
sed -i 's/^14/chr14/g' *bed
sed -i 's/^15/chr15/g' *bed
sed -i 's/^16/chr16/g' *bed
sed -i 's/^17/chr17/g' *bed
sed -i 's/^18/chr18/g' *bed
sed -i 's/^1/chr1/g' *bed
sed -i 's/^2/chr2/g' *bed
sed -i 's/^3/chr3/g' *bed
sed -i 's/^4/chr4/g' *bed
sed -i 's/^5/chr5/g' *bed
sed -i 's/^6/chr6/g' *bed
sed -i 's/^7/chr7/g' *bed
sed -i 's/^8/chr8/g' *bed
sed -i 's/^9/chr9/g' *bed
# 第三步（给bed文件添加表头）
#### 表头内容如下
track name="my_bed_file" description="This is a BED file"
# 第四步（将bed文件转为annovar输入文件）
perl /home/caominghao/software/21.ANNOVAR/annovar/convert2annovar.pl -format bed PI.bed > PI.avinput
# 注释
perl /home/caominghao/software/21.ANNOVAR/annovar/annotate_variation.pl -out PI -build Duroc PI.avinput /home/caominghao/software/21.ANNOVAR/annovar/Duroc/
```
## 5.提取基因
```shell
cat PI.variant_function | awk -F "\t" '{print $2}' | grep -v ^NONE | awk -F ',' 'BEGIN {OFS="\n"} {print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25,$26,$27,$28,$29,$30}' | sort | uniq | grep -v ^rna- | grep -v ^NONE | grep -v TRNA | awk -F "(" '{print $1}' | awk -F "-" '{print $2}' | sort | uniq | less -SN
```
