# gatk-snp-emmax
author: caominghao    date: 2024.6.26
****
## 1.准备文件
```shell
# 删除缺失值61m30.042s
time vcftools --gzvcf 493.snp.filter.pass.gt.vcf.gz --max-missing 0.1 --maf 0.05 --recode --stdout | bgzip > 493.snp.filter.pass.gt.noN.vcf.gz
time plink --vcf 493.snp.filter.pass.gt.noN.vcf.gz --make-bed --out 493.snp.filter.pass.gt.noN
cat 493.snp.filter.pass.gt.noN.bim | awk 'BEGIN {OFS="\t"} {print $1,$1"_"$4"_"$5"_"$6,$3,$4,$5,$6}' > 493.snp.filter.pass.gt.noN.bim2
mv 493.snp.filter.pass.gt.noN.bim2 493.snp.filter.pass.gt.noN.bim

# 耗时：1m15.872s
time plink --bfile 493.snp.filter.pass.gt \
--keep keep.list --make-bed \
--out 206
# 质控
# 1.使用两个指标：geno、maf质控(耗时： 1m55.716s)
time plink --bfile 206 \
--geno 0.01 --maf 0.1 \
--make-bed \
--out 206.qc 
# 2.使用LD指标：LD质控
time plink --bfile 206.qc \
--indep-pairwise 50 10 0.5 \
--out 206.qc.ld
time plink --bfile 206.qc \
--extract 206.qc.ld.prune.in \
--make-bed --out 206.qc.ld
```
## 2. 转为tfam
```shell
time plink --bfile 206.qc.ld --recode12 --transpose --out prefix2
```
## 3.emmax
```shell
time taskset -c 1 emmax-kin-intel64 -v -d 10 prefix2
# 使用emmax跑gwas
# 有协变量
time taskset -c 1 emmax-intel64 -v -d 10 -t prefix2 -p bh.yc -k prefix2.aBN.kinf -c q4.cov.txt -o bh.q4.cov
# 不用协变量
time taskset -c 1 emmax-intel64 -v -d 10 -t prefix2 -p bh.yc -k prefix2.aBN.kinf -o bh.q4.cov
# 看点

awk '$4<0.05/1351664 {print $1,$4}' bh.q4.cov.ps | grep ^9

# 画图
library(data.table)
setwd("/home/caominghao/02.HJB/06.pop_select/04.GWAS/")
p<-fread("bh.q4.cov.ps",data.table=F)
map<-read.table("206.qc.ld.bim")
pvalue<- cbind(map[,c(2,1,4)],p[,4])
colnames(pvalue)<-c("SNP","Chr","Pos","Pvalue")
library(CMplot)
CMplot(pvalue,plot.type="q",conf.int.col=NULL,box=TRUE,file="pdf",dpi=300,file.output=TRUE,verbose=TRUE)#QQ图
CMplot(pvalue, plot.type="m", LOG10=TRUE, ylim=NULL, threshold=c(0.05,1)/2912875,threshold.lty=c(1,2),threshold.lwd=c(1,1), threshold.col=c("black","grey"), amplify=TRUE,bin.size=1e6,col = c("#1B8D38","#F35B05"),signal.col=c("#212C71","#5C80CB"),signal.cex=c(1,1),signal.pch=c(19,19),file="pdf",dpi=300,file.output=TRUE,verbose=TRUE)



```
# 4.画图
```R
# 制备文件
cat bh.q4.cov.ps | awk 'BEGIN {OFS="\t"} {print $1,$4}' | awk -F '_' 'BEGIN {OFS="\t"} {print $1"_"$2"_"NR,$1,$2,$3}' | awk 'BEGIN {OFS="\t"} {print $1,$2,$3,$5}' > imf.gwas.txt
# 
library(CMplot)
setwd("/home/caominghao/02.HJB/06.pop_select/05.candid_gene/02.gwas_gene/")
getwd()
# gwas(曼哈顿)
colour = read.table("imf.gwas.txt", header = TRUE)
CMplot(colour,plot.type = "m",LOG10=TRUE,threshold=0.05/2912875,
       threshold.lty=2,threshold.col="red", amplify = TRUE, 
       file="pdf", dpi=300, cex=0.1,signal.cex=0.4,
       col = c("#6AEBDD","#EC7238"),
       file.output=TRUE, height=3,width=6,verbose=TRUE)
       colour = read.table("imf.gwas.txt", header = TRUE)
# gwas(qq plot)
CMplot(colour,plot.type="q",conf.int.col=NULL,box=TRUE,file="pdf",dpi=300,file.output=TRUE,verbose=TRUE)#QQ图

# 选择信号
fst_ele=read.table("chr2.fst",header = TRUE)
CMplot(fst_ele, plot.type="m",  LOG10=FALSE, ylab="Fst", ylim=c(0,1),amplify =TRUE,
       col = c("#8697CB","#EE797A"), band=0.5, threshold.lty=2, threshold.lwd=1, threshold.col="red", threshold=0.269774,
       cex=0.1, chr.den.col=NULL, height=3,width=6,signal.cex = 0.3,
       file="pdf", memo="13", dpi=300, file.output=TRUE, verbose=TRUE) 
```

# 4.1imf.gwas.txt文件格式
```shell
SNP Chr Pos snp_GL
1_5020 1 5020 0.585785724175595
1_10395 1 10395 0.218771719665273
1_13207 1 13207 0.287158063680811
1_30614 1 30614 0.488135939364738
1_40795 1 40795 0.127415784727471
1_44763 1 44763 0.0311007181182351
1_56810 1 56810 0.80744541833282
1_59270 1 59270 0.267094136576917
1_62716 1 62716 0.215934076669595
1_65354 1 65354 0.4660409783665
```






## 4.提取1%的点并且进行annovar的注释
```shell
## 在/home/caominghao/02.HJB/06.pop_select/05.candid_gene/02.gwas_gene/
#  转bed
awk '$4<0.05/1351664 {print $1,$4}' bh.q4.cov.ps | awk -F '_' '{print $1,$2,$3,$4}' | awk -F ' ' 'BEGIN {OFS="\t"} {print $1,$2,$2,$5}' > gwas.bed
cat gwas.bed | awk 'BEGIN {OFS="\t"} {print $1,$2-250000,$3+250000,$4}' > gwas.region.bed
# 注释
sed -i 's/^10/chr10/g' *bed
sed -i 's/^11/chr11/g' *bed
sed -i 's/^12/chr12/g' *bed
sed -i 's/^13/chr13/g' *bed
sed -i 's/^14/chr14/g' *bed
sed -i 's/^15/chr15/g' *bed
sed -i 's/^16/chr16/g' *bed
sed -i 's/^17/chr17/g' *bed
sed -i 's/^18/chr18/g' *bed
sed -i 's/^1/chr1/g' *bed
sed -i 's/^2/chr2/g' *bed
sed -i 's/^3/chr3/g' *bed
sed -i 's/^4/chr4/g' *bed
sed -i 's/^5/chr5/g' *bed
sed -i 's/^6/chr6/g' *bed
sed -i 's/^7/chr7/g' *bed
sed -i 's/^8/chr8/g' *bed
sed -i 's/^9/chr9/g' *bed
# 第三步（给bed文件添加表头）
#### 表头内容如下
track name="my_bed_file" description="This is a BED file"
# 第四步（将bed文件转为annovar输入文件）
perl /home/caominghao/software/21.ANNOVAR/annovar/convert2annovar.pl -format bed gwas.region.bed > gwas.avinput
# 注释
perl /home/caominghao/software/21.ANNOVAR/annovar/annotate_variation.pl -out gwas -build Duroc gwas.avinput /home/caominghao/software/21.ANNOVAR/annovar/Duroc/


## 提取基因
cat gwas.variant_function | awk -F "\t" '{print $2}' | grep -v ^NONE | awk -F ',' 'BEGIN {OFS="\n"} {print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25,$26,$27,$28,$29,$30}' | sort | uniq | grep -v ^rna- | grep -v ^NONE | grep -v TRNA | awk -F "(" '{print $1}' | awk -F "-" '{print $2}' | sort | uniq 
```

## 4.跑admixture

```shell
for i in {2..20}; do jsub -R "span[hosts=1]" -q normal -n 64 -M 500000000 -J admixture.${i} \
-e /storage/public/home/2021055373/01.Data/01.HJB/02.admixture-gwas/log/admixture.${i}.erro \
-o /storage/public/home/2021055373/01.Data/01.HJB/02.admixture-gwas/log/admixture.${i}.ok \
"/storage/public/home/2021055373/software/dist/admixture_linux-1.3.0/admixture --cv \
/storage/public/home/2021055373/01.Data/01.HJB/02.admixture-gwas/206.qc.ld.bed ${i} | \
tee /storage/public/home/2021055373/01.Data/01.HJB/02.admixture-gwas/log${i}.out"; done

for i in {2..10};do time admixture -j30 --cv 206.qc.ld.bed ${i} | tee log${i}.out & done
# 合成协变量
awk -F ' ' '{print $1,$2,1}' prefix2.tfam > 1
awk -F ' ' '{print $1,$2,$3,$4,$5,$6,$7,$8}' ./03.admixture/206.qc.ld.8.Q > 2
paste -d ' ' 1 2 > q4.cov.txt
```

## 5.LDBlockShow

```shell
# 提取
time bcftools view -r 9:118000000-118900000 HJB.CommonChr.SNP.noallele.SNPagain.renameChr.renameSample.keep.raw.vcf.gz -Oz -o PAPPA2.vcf.gz
# 画图
time LDBlockShow -SeleVar 1 \
-InPlink HJB.CommonChr.snp.noallele.snpagain.renameChr.keep.renameSample.raw.keep.qc.ld \
-OutPut PAPPA2-small4 \
-Region 9:118000000-118900000 \
-MAF 0.1 \
-InGWAS gwas \
-InGFF Duroc.gff

time LDBlockShow -InVCF PAPPA2.vcf.gz -OutPut PAPPA2 -InGWAS gwas.pvalue -InGFF In.gff -Region 9:118000000-118900000 -OutPng -SeleVar 2
```
## 6.画图
```shell

```
