i# gatk-snp-pi
author: caominghao  date: 2024.6.24
****
## 1.每个群体1-7个个体，计算pi值
```shell
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr1.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome1/${i}.Chr1.pi
done
# chromosome2
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr2.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome2/${i}.Chr2.pi
done
# chromosome3
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr3.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome3/${i}.Chr3.pi
done
# chromosome4
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr4.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome4/${i}.Chr4.pi
done
# chromosome5
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr5.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome5/${i}.Chr5.pi
done
# chromosome6
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr6.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome6/${i}.Chr6.pi
done
# chromosome7
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr7.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome7/${i}.Chr7.pi
done
# chromosome8
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr8.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome8/${i}.Chr8.pi
done
# chromosome9
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr9.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome9/${i}.Chr9.pi
done
# chromosome10
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr10.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome10/${i}.Chr10.pi
done
# chromosome11
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr11.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome11/${i}.Chr11.pi
done
# chromosome12
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr12.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome12/${i}.Chr12.pi
done
# chromosome13
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr13.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome13/${i}.Chr13.pi
done
# chromosome14
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr14.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome14/${i}.Chr14.pi
done
# chromosome15
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr15.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome15/${i}.Chr15.pi
done
# chromosome16
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr16.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome16/${i}.Chr16.pi
done
# chromosome17
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr17.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome17/${i}.Chr17.pi
done
# chromosome18
for i in `cat list`;do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr18.vcf.gz --keep ${i}.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome18/${i}.Chr18.pi
done
```
## 2.单独计算W和SMT猪的PI
```shell
# warthog
for i in {1..18};do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr${i}.vcf.gz --keep W.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome${i}/W.Chr${i}.pi & 
done
# sumatra
for i in {1..18};do 
vcftools --gzvcf /home/lidong/Data/05.HJB_vcf/01.vcf/493.snp.filter.pass.gt.Chr${i}.vcf.gz --keep SMT.list --window-pi 1000 --window-pi-step 1000 --max-missing 0.1 --maf 0.01 --out ./Chromosome${i}/SMT.Chr${i}.pi & 
done

for j in {1..18};do 
for i in W ;do cat ./Chromosome${j}/${i}.Chr${j}.pi.windowed.pi | awk '{sum += $NF} END {print "'"Chromosome${j}.${i}"' average : ", sum/NR}';done;done
```



## 2.提取PI值

```shell
for j in {1..18};do 
for i in `cat list`;do cat ./Chromosome${j}/${i}.Chr${j}.pi.windowed.pi | awk '{sum += $NF} END {print "'"Chromosome${j}.${i}"' average : ", sum/NR}';done;done
```
